File: Kun/leetcode/19-Remove-Nth-Node-From-End-of-List/19-Remove-Nth-Node-From-End-of-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import ListNode

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummyHead = ListNode(next = head)
        fast = slow = dummyHead
        for _ in range(n-1):
            fast = fast.next
        if not fast:
            return dummyHead.next
        pre = dummyHead
        while fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next
        pre.next = slow.next
        return dummyHead.next

File: Kun/leetcode/19-Remove-Nth-Node-From-End-of-List/19-Remove-Nth-Node-From-End-of-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null) return null;
        ListNode dummyHead = new ListNode();
        dummyHead.next = head;
        int length = 1;
        ListNode nNode = dummyHead;
        while(nNode.next != null && length < n+1){
            nNode = nNode.next;
            length++;
        }
        if(length != n+1) return head;
        ListNode pre = dummyHead;
        ListNode cur = head;
        while(nNode.next != null){
            pre = cur;
            cur = cur.next;
            nNode = nNode.next;
        }
        pre.next = cur.next;
        return dummyHead.next;
    }
}